README.md
FastSocket
===============

Description
---------------

A high-throughput, synchronous Objective-C wrapper around BSD sockets for iOS and macOS.
Send and receive raw bytes over a socket as fast as possible. Includes methods
for transferring files while optionally computing a checksum for verification.
Swift compatible.

Use this class if fast network communication is what you need. If you want to
do something else while your network operations finish, then an asynchronous
API might be better.

For more information, please visit the [project homepage](http://github.com/dreese/FastSocket).
FastSocket is also available as a [CocoaPod](http://cocoapods.org/?q=fastsocket).

Download
---------------

Download the [latest release](https://github.com/dreese/FastSocket/releases) of FastSocket or try the [nightly version](https://github.com/dreese/FastSocket/archive/master.zip).

Examples
---------------

Create and connect a client socket.

	FastSocket *client = [[FastSocket alloc] initWithHost:@"localhost" andPort:@"34567"];
	[client connect];

Send a file.

	long sent = [client sendFile:@"/tmp/filetosend.txt"];

Receive a file of a given length.

	long received = [client receiveFile:@"/tmp/newlyreceivedfile.txt" length:1024];

Send a string.

	NSData *data = [@"test" dataUsingEncoding:NSUTF8StringEncoding];
	long count = [client sendBytes:[data bytes] count:[data length]];

Receive a string.

	char bytes[expectedLength];
	[client receiveBytes:bytes count:expectedLength];
	NSString *received = [[NSString alloc] initWithBytes:bytes length:expectedLength encoding:NSUTF8StringEncoding];

Send raw bytes.

	char data[] = {42};
	long sent = [client sendBytes:data count:1];

Receive available raw bytes up to the given limit.

	char data[42];
	long received = [client receiveBytes:data limit:42];

Receive the exact number of raw bytes given.

	char data[1000];
	long received = [client receiveBytes:data count:1000];

Close the connection.

	[client close];

Please check out the unit tests for more examples of how to use these classes.

Creator
---------------

[Daniel Reese](http://www.danielreese.com/)
[@dreese](http://twitter.com/dreese)

License
---------------

FastSocket is available under the [MIT license](http://opensource.org/licenses/MIT).


CHANGELOG.md
FastSocket
===============

Release Notes
---------------
2017 Nov 01 — [v1.6](https://github.com/dreese/FastSocket/releases/tag/v1.6)

	• Fixed a long-standing bug where the internal socket descriptor could become closed but not zeroed out.

2017 Oct 28 — [v1.5](https://github.com/dreese/FastSocket/releases/tag/v1.5)

	• Annotated code to improve auto-generated Swift interface.
	• Fixed several documentation issues.
	• Added document for contributions.

2017 Oct 06 — [v1.4](https://github.com/dreese/FastSocket/releases/tag/v1.4)

	• Changed -[FaskSocket timeout] and -[FaskSocket setTimeout:] methods so that the timeout value is a float, in order to handle sub-second values.
	• Fixed a compatibility issue with Xcode 9.
	• Added several unit tests.

2015 Jan 27 — [v1.3](https://github.com/dreese/FastSocket/releases/tag/v1.3)

	• Changed -[FaskSocket sendBytes:count:] method to return the actual number of bytes received instead of a BOOL. Now it matches the Readme.
	• Fixed a compiler warning caused by returning NO instead of nil from one of the init methods.
	• Added several unit tests.

2014 Feb 03 — [v1.2](https://github.com/dreese/FastSocket/releases/tag/v1.2)

	• Added -[FastSocket connect:] method for specifying a connection timeout, which is separate from the read/write timeout.
	• Added CocoaPod support with new podspec file.

2013 Oct 03 — [v1.1](https://github.com/dreese/FastSocket/releases/tag/v1.1)

	• Converted to ARC.
	• Added -[FastSocket isConnected] method.
	• Added -[FastSocket receiveBytes:count:] method for receiving an exact number of bytes. This differs from -[FastSocket receiveBytes:limit:] in that the new method waits for the given number of bytes is received, or a timeout, before returning.
	• Added header documentation for use in Xcode 5.

2012 Jun 24 — [v1.0](https://github.com/dreese/FastSocket/releases/tag/v1.0)

	• Initial release.


FastSocket.h
//
//  FastSocket.h
//  Copyright (c) 2011-2013 Daniel Reese <dan@danandcheryl.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import <Foundation/Foundation.h>


#define NEW_ERROR(num, str) [[NSError alloc] initWithDomain:@"FastSocketErrorDomain" code:(num) userInfo:[NSDictionary dictionaryWithObject:[NSString stringWithFormat:@"%s", (str)] forKey:NSLocalizedDescriptionKey]]

NS_ASSUME_NONNULL_BEGIN
@interface FastSocket : NSObject

#pragma mark - Properties

/**
 The file descriptor used to communicate to the remote machine.
 */
@property (nonatomic, readonly) int sockfd;

/**
 The host name of the remote machine.
 */
@property (nonatomic, readonly) NSString *host;

/**
 The port number of the remote machine.
 */
@property (nonatomic, readonly) NSString *port;

/**
 The last error that occured. This value is not set to nil after a successful call, so it is not
 appropriate to test this value to check for error conditions. Check for a NO or nil return value.
 */
@property (nullable, nonatomic, readonly) NSError *lastError;

#pragma mark - Initializers

/**
 Returns an initialized FastSocket object configured to connect to the given host name and port number.
 
 @param host The host name of the remote host.
 @param port The port number on which to connect.
 @return An initialized FastSocket object configured to connect to the given host name and port number.
 */
- (nullable instancetype)initWithHost:(NSString *)host andPort:(NSString *)port;

/**
 Returns an initialized FastSocket object configured to communicate throught the given file descriptor.
 This method is used primarily by a server socket to receive an incoming connection.
 
 @param fd The file descriptor to use for communication.
 @return An initialized FastSocket object configured to communicate throught the given file descriptor.
 */
- (nullable instancetype)initWithFileDescriptor:(int)fd;

/**
 Retrieves the internal buffer and its size for use outside the class. This buffer is good
 to use for sending and receiving bytes because it is a multiple of the segment size and
 allocated so that it is aligned on a memory page boundary.
 */
- (void)buffer:(void * _Nonnull * _Nonnull)buf size:(long *)size;

#pragma mark - Actions

/**
 Connect the socket to the remote host.
 
 @return YES if the connection succeeded, NO otherwise.
 */
- (BOOL)connect;

/**
 Connect the socket to the remote host with the given timeout value.
 
 @param timeout The maximum amount of time to wait for a connection to succeed.
 @return YES if the connection succeeded, NO otherwise.
 */
- (BOOL)connect:(long)timeout;

/**
 Returns whether the socket is currently connected.
 
 @return YES if the socket is connected, NO otherwise.
 */
- (BOOL)isConnected;

/**
 Closes the connection to the remote host.
 
 @return YES if the close succeeded, NO otherwise.
 */
- (BOOL)close;

/**
 Sends the specified number bytes from the given buffer.
 
 @param buf   The buffer containing the bytes to send.
 @param count The number of bytes to send, typically the size of the buffer.
 @return The actual number of bytes sent.
 */
- (long)sendBytes:(const void *)buf count:(long)count;

/**
 Receives an unpredictable number bytes up to the specified limit. Stores the bytes
 in the given buffer and returns the actual number received.
 
 @param buf   The buffer in which to store the bytes received.
 @param limit The maximum number of bytes to receive, typically the size of the buffer.
 @return The actual number of bytes received.
 */
- (long)receiveBytes:(void *)buf limit:(long)limit;

/**
 Receives the exact number of bytes specified unless a timeout or other error occurs.
 Stores the bytes in the given buffer and returns the actual number received. To check
 whether the correct number of bytes was received, compare the return value with the
 count specified.
 
 @param buf   The buffer in which to store the bytes received.
 @param count The exact number of bytes to receive, typically the size of the buffer.
 @return The actual number of bytes received.
 */
- (long)receiveBytes:(void *)buf count:(long)count;

/**
 Sends the contents of the file at the specified path. Uses an internal buffer to read
 a block of data from the file and send it over the network.
 
 @param path The path of the file to send.
 @return The actual number of bytes sent.
 */
- (long)sendFile:(NSString *)path;

/**
 Receives a file with the given length and writes it to the specified path. Uses an
 internal buffer to receive a block of data from the network and write it to disk.
 Overwrites any existing file.
 
 @param path   The path of the file to create.
 @param length The length of the file, measured in bytes.
 @return The actual number of bytes received.
 */
- (long)receiveFile:(NSString *)path length:(long)length;

/**
 Receives a file with the given length and writes it to the specified path and calculates
 its MD5 hash. The hash can be used for error checking. Uses an internal buffer to receive
 a block of data from the network and write it to disk. Overwrites any existing file.
 
 @param path   The path of the file to create.
 @param length The length of the file, measured in bytes.
 @param hash   An optional data object in which to store the calculated MD5 hash of the file.
 @return The actual number of bytes received.
 */
- (long)receiveFile:(NSString *)path length:(long)length md5:(NSData * _Nonnull * _Nullable)hash;

#pragma mark - Settings

/**
 Returns the number of seconds to wait without any network activity before giving up and
 returning an error. The default is zero seconds, in which case it will never time out.
 
 @return The current timeout value in seconds.
 */
- (float)timeout;

/**
 Sets the number of seconds to wait without any network activity before giving up and
 returning an error. The default is zero seconds, in which case it will never time out.
 
 @param seconds The number of seconds to wait before timing out.
 @return YES if the timeout value was set successfully, NO otherwise.
 */
- (BOOL)setTimeout:(float)seconds;

/**
 Returns the maximum segment size. The segment size is the largest amount of
 data that can be transmitted within a single packet. Too large of a value may result in packets
 being broken apart and reassembled during transmission, which is normal but can slow things
 down. Too small of a value may result in lots of overhead, which can also slow things down.
 
 A default value is automatically negotiated when a connection is established. However, the
 negotiation process may incorporate assumptions that are incorrect. If you understand the
 network condidtions, setting this value correctly may increase performance.
 
 @return The current maximum segment value, measured in bytes.
 */
- (int)segmentSize;

/**
 Sets the maximum segment size. The segment size is the largest amount of
 data that can be transmitted within a single packet, excluding headers. Too large of a value
 may result in packets being broken apart and reassembled during transmission, which is normal
 but can slow things down. Too small of a value may result in lots of overhead, which can
 also slow things down.
 
 A default value is automatically negotiated when a connection is established. However, the
 negotiation process may incorporate assumptions that are incorrect. If you understand the
 network condidtions, setting this value correctly may increase performance.
 
 @param bytes The maximum segment size, measured in bytes.
 @return YES if the segment size value was set successfully, NO otherwise.
 */
- (BOOL)setSegmentSize:(int)bytes;

@end
NS_ASSUME_NONNULL_END


FastServerSocket.h
//
//  FastServerSocket.h
//  Copyright (c) 2011-2013 Daniel Reese <dan@danandcheryl.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import <Foundation/Foundation.h>

@class FastSocket;
@interface FastServerSocket : NSObject {
@protected
	long timeout;
}

#pragma mark - Properties

/**
 The file descriptor used to communicate to the remote machine.
 */
@property (nonatomic, readonly) int sockfd;

/**
 The port number of the local machine on which to listen for incoming connections.
 */
@property (nonatomic, readonly) NSString *port;

/**
 The last error that occured. This value is not set to nil after a successful call, so it is not
 appropriate to test this value to check for error conditions. Check for a NO or nil return value.
 */
@property (nonatomic, readonly) NSError *lastError;

#pragma mark - Initializers

/**
 Returns an initialized FastSocketSocket object configured to connect to the given host name and port number.
 
 @param port The port number on which to connect.
 @return An initialized FastSocketSocket object configured to connect to the given host name and port number.
 */
- (id)initWithPort:(NSString *)port __attribute__((nonnull));

/**
 Returns an initialized FastSocketSocket object configured to communicate throught the given file descriptor.
 This method is primary used by a server socket to receive an incoming connection.
 
 @param fd The file descriptor to use for communication.
 @return An initialized FastSocketSocket object configured to communicate throught the given file descriptor.
 */
- (id)initWithFileDescriptor:(int)fd;

#pragma mark Actions

/**
 Starts listening for incoming connections.
 
 @return YES if the listen succeeded, NO otherwise.
 */
- (BOOL)listen;

/**
 Accepts an incoming connection from a remote host. Blocks until a connection is received.
 
 @return A socket connected to the remote host.
 */
- (FastSocket *)accept;

/**
 Closes the connection to the remote host.
 
 @return YES if the close succeeded, NO otherwise.
 */
- (BOOL)close;

#pragma mark Settings

/**
 Returns the number of seconds to wait without any network activity before giving up and
 returning an error. The default is zero seconds, in which case it will never time out.
 
 @return The current timeout value in seconds.
 */
- (long)timeout;

/**
 Sets the number of seconds to wait without any network activity before giving up and
 returning an error. The default is zero seconds, in which case it will never time out.
 
 @param seconds The number of seconds to wait before timing out.
 @return YES if the timeout value was set successfully, NO otherwise.
 */
- (BOOL)setTimeout:(long)seconds;

@end


FastSocketTest.m
//
//  FastSocketTest.h
//  Copyright (c) 2011-2013 Daniel Reese <dan@danandcheryl.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import "FastSocket.h"
#import "FastServerSocket.h"
#import <XCTest/XCTest.h>


@interface FastSocketTest : XCTestCase {
	FastSocket *client;
	FastServerSocket *server;
}

@end


@implementation FastSocketTest

- (void)setUp {
	server = [[FastServerSocket alloc] initWithPort:@"34567"];
	client = [[FastSocket alloc] initWithHost:@"localhost" andPort:@"34567"];
}

- (void)tearDown {
	[client close];
	[server close];
}

#pragma mark Tests

- (void)testConnect {
	// No failures yet.
	XCTAssertNil([client lastError]);
	
	// Nothing is listening yet.
	XCTAssertFalse([client connect]);
	XCTAssertNotNil([client lastError]);
	
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Connection should now succeed.
	XCTAssertTrue([client connect]);
	//STAssertNil([client lastError], @"Last error should be nil"); // TODO: Not sure if this should be cleared out or just left alone.
}

- (void)testDoubleCloseAfterConnectionFailure {
    XCTAssertFalse([client connect:0]);
    XCTAssertTrue([client close]);
    XCTAssertTrue([client close]);
    
    XCTAssertFalse([client connect:1]);
    XCTAssertTrue([client close]);
    XCTAssertTrue([client close]);
}

- (void)testConnectWithDefaultTimeout {
	// Connect to a non-routable IP address. See https://stackoverflow.com/a/40459270
	[client close];
	client = [[FastSocket alloc] initWithHost:@"example.com" andPort:@"81"];
	
	// Connection should timeout.
	NSTimeInterval startTime = [NSDate timeIntervalSinceReferenceDate];
	XCTAssertFalse([client connect]);
	NSTimeInterval endTime = [NSDate timeIntervalSinceReferenceDate];
	XCTAssertNotNil([client lastError]);
	
	// Default timeout is 75 seconds on my machine.
	NSTimeInterval actualTime = endTime - startTime;
	XCTAssertTrue(actualTime >= 75.0, @"timeout was %.2f", actualTime);
	XCTAssertTrue(actualTime < 76.3, @"timeout was %.2f", actualTime);
}

- (void)testConnectWithCustomTimeout {
	// Connect to a non-routable IP address. See https://stackoverflow.com/a/40459270
	[client close];
	client = [[FastSocket alloc] initWithHost:@"example.com" andPort:@"81"];
	
	// Connection should timeout.
	NSTimeInterval startTime = [NSDate timeIntervalSinceReferenceDate];
	XCTAssertFalse([client connect:10]);
	NSTimeInterval endTime = [NSDate timeIntervalSinceReferenceDate];
	XCTAssertNotNil([client lastError]);
	
	// Check the duration of the timeout.
	NSTimeInterval actualTime = endTime - startTime;
	XCTAssertTrue(actualTime >= 9.9, @"timeout was %.2f", actualTime);
	XCTAssertTrue(actualTime < 10.1, @"timeout was %.2f", actualTime);
}

- (void)testIsConnected {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Check connected state before and after connecting, and after closing.
	XCTAssertFalse([client isConnected]);
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client isConnected]);
	XCTAssertTrue([client close]);
	XCTAssertFalse([client isConnected]);
	
	// Close server socket and verify that the client knows it's no longer connected.
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client isConnected]);
	XCTAssertTrue([server close]);
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.2]];
	XCTAssertFalse([client isConnected]);
}

- (void)testIsConnectedAfterRemoteClose {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(listentAndClose:) toTarget:self withObject:@2];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Check connected state before and after connecting, and after closing.
	XCTAssertFalse([client isConnected]);
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client isConnected]);
	
	// Wait for server to close.
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:3]];
	NSLog(@"checking connected status");
	XCTAssertFalse([client isConnected]);
}

- (void)testTimeoutBefore {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value before connect.
	XCTAssertTrue([client setTimeout:100]);
	XCTAssertTrue([client connect]);
	XCTAssertEqual([client timeout], 100);
	XCTAssertTrue([client close]);
}

- (void)testTimeoutAfter {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value after connect.
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client setTimeout:100]);
	XCTAssertEqual([client timeout], 100);
	XCTAssertTrue([client close]);
}

- (void)testTimeoutMultipleBefore {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value twice before connect.
	XCTAssertTrue([client setTimeout:100]);
	XCTAssertTrue([client setTimeout:101]);
	XCTAssertTrue([client connect]);
	XCTAssertEqual([client timeout], 101);
	XCTAssertTrue([client close]);
}

- (void)testTimeoutMultipleAfter {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value twice after connect.
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client setTimeout:100]);
	XCTAssertTrue([client setTimeout:101]);
	XCTAssertEqual([client timeout], 101);
	XCTAssertTrue([client close]);
}

- (void)testTimeoutSubsecondBefore {
    // Spawn server thread.
    [NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
    [NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    
    // Set value before connect.
    float timeout = arc4random_uniform(1000) / 1000.0f;
    XCTAssertTrue([client setTimeout:timeout]);
    XCTAssertTrue([client connect]);
    XCTAssertEqual([client timeout], timeout);
    XCTAssertTrue([client close]);
}

- (void)testTimeoutSubsecondAfter {
    // Spawn server thread.
    [NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
    [NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    
    // Set value after connect.
    float timeout = arc4random_uniform(1000) / 1000.0f;
    XCTAssertTrue([client connect]);
    XCTAssertTrue([client setTimeout:timeout]);
    XCTAssertEqual([client timeout], timeout);
    XCTAssertTrue([client close]);
}

- (void)testClearTimeout {
    // Spawn server thread.
    [NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
    [NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    
    // Set value before connect.
    XCTAssertTrue([client connect]);
    XCTAssertEqual([client timeout], 0.0f);
    XCTAssertTrue([client setTimeout:100]);
    XCTAssertTrue([client setTimeout:0]);
    XCTAssertEqual([client timeout], 0.0f);
    XCTAssertTrue([client close]);
}

- (void)testSegmentSizeBefore {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value before connect.
	XCTAssertTrue([client setSegmentSize:10000]);
	XCTAssertTrue([client connect]);
	XCTAssertEqual([client segmentSize], 10000);
	XCTAssertTrue([client close]);
}

- (void)testSegmentSizeAfter {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value after connect.
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client setSegmentSize:10000]);
	XCTAssertEqual([client segmentSize], 10000);
	XCTAssertTrue([client close]);
}

- (void)testSegmentSizeMultipleBefore {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value twice before connect.
	XCTAssertTrue([client setSegmentSize:10000]);
	XCTAssertTrue([client setSegmentSize:10001]);
	XCTAssertTrue([client connect]);
	XCTAssertEqual([client segmentSize], 10001);
	XCTAssertTrue([client close]);
}

- (void)testSegmentSizeMultipleAfter {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(simpleListen:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Set value twice after connect.
	XCTAssertTrue([client connect]);
	XCTAssertTrue([client setSegmentSize:10000]);
	XCTAssertFalse([client setSegmentSize:10001]);
	XCTAssertTrue([client setSegmentSize:9999]);
	XCTAssertEqual([client segmentSize], 9999);
	XCTAssertTrue([client close]);
}

- (void)testSendingAndReceivingBytes {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(listenAndRepeat:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Send byte array.
	long len = 10;
	unsigned char sent[] = {1, -2, 3, -4, 5, -6, 7, -8, 9, 0};
	[client connect];
	XCTAssertEqual([client sendBytes:sent count:len], len, @"send error: %@", [client lastError]);
	[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Receive byte array.
	unsigned char received[len];
	XCTAssertEqual([client receiveBytes:received count:len], len, @"receive error: %@", [client lastError]);
	
	// Compare results.
	XCTAssertEqual(memcmp(sent, received, len), 0);
}

- (void)testSendingAndReceivingRandomBytes {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(listenAndRepeat:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Create a random byte array larger than the buffer.
	long len = 1024 * 10 * 200;
	unsigned char sent[len];
	for (int i = 0; i < len; ++i) {
		sent[i] = (unsigned char)(random() % 256);
	}
	NSLog(@"sending %li bytes", len);
	
	// Send the array.
	[client connect];
	XCTAssertEqual([client sendBytes:sent count:len], len, @"send error: %@", [client lastError]);
	[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Receive the array.
	unsigned char received[len];
	XCTAssertEqual([client receiveBytes:received count:len], len, @"receive error: %@", [client lastError]);
	
	// Compare results.
	XCTAssertEqual(memcmp(sent, received, len), 0);
}

- (void)testSendingAndReceivingStrings {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(listenAndRepeat:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	[client connect];
	
	// Send the string.
	NSString *original = @"This is å striñg to tëst séndîng. 😎";
	NSData *data = [original dataUsingEncoding:NSUTF8StringEncoding];
	long len = [data length];
	XCTAssertEqual([client sendBytes:[data bytes] count:len], len, @"send error: %@", [client lastError]);
	[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Receive the string.
	char bytes[len];
	XCTAssertEqual([client receiveBytes:bytes count:len], len, @"receive error: %@", [client lastError]);
	NSString *received = [[NSString alloc] initWithBytes:bytes length:len encoding:NSUTF8StringEncoding];
	
	// Compare results.
	XCTAssertEqualObjects(received, original);
}

- (void)testReceiveBytesWithDelay {
	int count = 10;
	
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(sendWithDelay:) toTarget:self withObject:@(count)];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Receive the array.
	[client connect];
	unsigned char received[count];
	memset(received, 0, count);
	XCTAssertEqual([client receiveBytes:received count:count], count, @"receive error: %@", [client lastError]);
	XCTAssertEqual(received[count - 1], count - 1, @"incorrect result");
}

- (void)testReceiveBytesWithTimeout {
	// Spawn server thread.
	[NSThread detachNewThreadSelector:@selector(listenAndRepeat:) toTarget:self withObject:nil];
	[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Send byte array.
	long len = 10;
	unsigned char sent[len];
	memset(sent, 1, len);
	[client connect];
	XCTAssertEqual([client sendBytes:sent count:len], len, @"send error: %@", [client lastError]);
	[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
	
	// Receive byte array.
	unsigned char received[len + 1];
	[client setTimeout:3];
	
	NSTimeInterval startTime = [NSDate timeIntervalSinceReferenceDate];
	XCTAssertEqual([client receiveBytes:received count:(len + 1)], len, @"receive error: %@", [client lastError]);
	NSTimeInterval endTime = [NSDate timeIntervalSinceReferenceDate];
	XCTAssertNotNil([client lastError]);
	
	// Check the duration of the timeout.
	NSTimeInterval actualTime = endTime - startTime;
	XCTAssertTrue(actualTime >= 2.9, @"timeout was %.2f", actualTime);
	XCTAssertTrue(actualTime < 3.1, @"timeout was %.2f", actualTime);
	
	// Compare results.
	XCTAssertEqual(memcmp(sent, received, len), 0);
}

- (void)testChecksum {
    int count = 10;
    
    // Spawn server thread.
    [NSThread detachNewThreadSelector:@selector(sendWithDelay:) toTarget:self withObject:@(count)];
    [NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
    
    // Receive the array.
    [client connect];
    unsigned char received[count];
    memset(received, 0, count);
    
    NSData *checksum = nil;
    long bytesReceived = [client receiveFile:@"/tmp/test.txt" length:count md5:&checksum];
    XCTAssertEqual(bytesReceived, count, @"receive error: %@", [client lastError]);
    
    NSMutableString *checksumString = [NSMutableString stringWithCapacity:checksum.length * 2];
    const unsigned char *bytes = checksum.bytes;
    for (NSUInteger i = 0; i < checksum.length; ++i) {
        [checksumString appendFormat:@"%02x", (unsigned int)bytes[i]];
    }
    XCTAssertEqualObjects(checksumString, @"c56bd5480f6e5413cb62a0ad9666613a", @"incorrect checksum");
}

#pragma mark Helpers

- (void)simpleListen:(id)obj {
	@autoreleasepool {
		[server listen]; // Incoming connections just queue up.
	}
}

- (void)listentAndClose:(NSNumber *)delay {
	@autoreleasepool {
		NSLog(@"started listening");
		[server listen];
		
		// Wait before closing.
		[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:delay.doubleValue]];
		NSLog(@"closing connection");
		[server close];
	}
}

- (void)listenAndRepeat:(id)obj {
	@autoreleasepool {
		NSLog(@"started listening");
		[server listen];
		
		FastSocket *incoming = [server accept];
		if (!incoming) {
			NSLog(@"accept error: %@", [server lastError]);
			return;
		}
		
		// Read some bytes then echo them back.
		int bufSize = 2048;
		unsigned char buf[bufSize];
		long count = 0;
		do {
			// Read bytes.
			count = [incoming receiveBytes:buf limit:bufSize];
			
			// Write bytes.
			long remaining = count;
			while (remaining > 0) {
				count = [incoming sendBytes:buf count:remaining];
				remaining -= count;
			}
			
			// Allow other threads to work.
			[[NSRunLoop currentRunLoop] runUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.1]];
		} while (count > 0);
		
		XCTAssertEqual(count, 0, @"error: %@", [incoming lastError]);
	}
}

- (void)sendWithDelay:(NSNumber *)count {
	@autoreleasepool {
		NSLog(@"started listening");
		[server listen];
		
		FastSocket *incoming = [server accept];
		if (!incoming) {
			NSLog(@"accept error: %@", [server lastError]);
			return;
		}
		
		// Send each byte 0.5 sec apart.
		long send = 0;
		unsigned char buf[1];
		for (char i = 0; i < count.intValue; ++i) {
			buf[0] = i;
			send += [incoming sendBytes:buf count:1];
			
			[NSThread sleepUntilDate:[NSDate dateWithTimeIntervalSinceNow:0.5]];
		}
		
		XCTAssertEqual(send, 0, @"error: %@", [incoming lastError]);
	}
}

@end
